Compile error under 2.0.8 if library function explicitly returns const __FlashStringHelper * from F() macro

### Board

ESP32

### Device Description

YUBOX-Node

### Hardware Configuration

No special hardware configuration needed.

### Version

v2.0.8

### IDE Name

Arduino IDE 2.1.0

### Operating System

Fedora Linux 38 x86_64

### Flash frequency

80MHz

### PSRAM enabled

no

### Upload speed

115200

### Description

After updating to Arduino-ESP32 2.0.8, the attached sketch will no longer compile, and displays the following error messages:
```
In file included from /home/alex/.arduino15/packages/esp32-alternatives/2.0.8/hardware/esp32/2.0.8/cores/esp32/Arduino.h:176,
                 from /home/alex/programa/practica/arduino/esp32/test-bug-compile-fail-flashstringhelper/test-bug-compile-fail-flashstringhelper.ino:1:
/home/alex/programa/practica/arduino/esp32/test-bug-compile-fail-flashstringhelper/test-bug-compile-fail-flashstringhelper.ino: In function 'const __FlashStringHelper* testfunc()':
/home/alex/.arduino15/packages/esp32-alternatives/2.0.8/hardware/esp32/2.0.8/cores/esp32/WString.h:38:42: error: cannot convert 'const char*' to 'const __FlashStringHelper*' in return
 #define F(string_literal) (string_literal)
                                          ^
/home/alex/programa/practica/arduino/esp32/test-bug-compile-fail-flashstringhelper/test-bug-compile-fail-flashstringhelper.ino:5:10: note: in expansion of macro 'F'
   return F("test");
          ^

exit status 1

Compilation error: exit status 1
```

The same sketch compiles correctly under previous 2.0.7.

### Sketch

```cpp
#include <Arduino.h>

const __FlashStringHelper * testfunc(void)
{
  return F("test");
}

void setup() {
  // put your setup code here, to run once:

}

void loop() {
  // put your main code here, to run repeatedly:
  delay(1000);
}
```


### Debug Message

```plain
In file included from /home/alex/.arduino15/packages/esp32-alternatives/2.0.8/hardware/esp32/2.0.8/cores/esp32/Arduino.h:176,
                 from /home/alex/programa/practica/arduino/esp32/test-bug-compile-fail-flashstringhelper/test-bug-compile-fail-flashstringhelper.ino:1:
/home/alex/programa/practica/arduino/esp32/test-bug-compile-fail-flashstringhelper/test-bug-compile-fail-flashstringhelper.ino: In function 'const __FlashStringHelper* testfunc()':
/home/alex/.arduino15/packages/esp32-alternatives/2.0.8/hardware/esp32/2.0.8/cores/esp32/WString.h:38:42: error: cannot convert 'const char*' to 'const __FlashStringHelper*' in return
 #define F(string_literal) (string_literal)
                                          ^
/home/alex/programa/practica/arduino/esp32/test-bug-compile-fail-flashstringhelper/test-bug-compile-fail-flashstringhelper.ino:5:10: note: in expansion of macro 'F'
   return F("test");
          ^

exit status 1

Compilation error: exit status 1
```


### Other Steps to Reproduce

_No response_

### I have checked existing issues, online documentation and the Troubleshooting Guide

- [X] I confirm I have checked existing issues, online documentation and Troubleshooting guide.

Same problem here with the WS2812FX Library too.

I```
n file included from C:\Users\ryanc\AppData\Local\Arduino15\packages\esp32\hardware\esp32\2.0.8\cores\esp32/Arduino.h:176,
                 from c:\Users\ryanc\Documents\Arduino\libraries\Adafruit_NeoPixel/Adafruit_NeoPixel.h:41,
                 from c:\Users\ryanc\Documents\Arduino\libraries\WS2812FX\src\WS2812FX.h:43,
                 from c:\Users\ryanc\Documents\Arduino\libraries\WS2812FX\src\WS2812FX.cpp:55:
c:\Users\ryanc\Documents\Arduino\libraries\WS2812FX\src\WS2812FX.cpp: In member function 'const __FlashStringHelper* WS2812FX::getModeName(uint8_t)':
C:\Users\ryanc\AppData\Local\Arduino15\packages\esp32\hardware\esp32\2.0.8\cores\esp32/WString.h:38:42: error: cannot convert 'const char*' to 'const __FlashStringHelper*' in return
 #define F(string_literal) (string_literal)
                                          ^
c:\Users\ryanc\Documents\Arduino\libraries\WS2812FX\src\WS2812FX.cpp:404:12: note: in expansion of macro 'F'
     return F("");
            ^
Multiple libraries were found for "WiFi.h"
  Used: C:\Users\ryanc\AppData\Local\Arduino15\packages\esp32\hardware\esp32\2.0.8\libraries\WiFi
  Not used: C:\Users\ryanc\Documents\Arduino\libraries\WiFiEspAT
  Not used: C:\Users\ryanc\Documents\Arduino\libraries\WiFiNINA
exit status 1

Compilation error: exit status 1
```

--- a/cores/esp32/WString.h
+++ b/cores/esp32/WString.h
@@ -34,8 +34,8 @@
 // A pure abstract class forward used as a means to proide a unique pointer type
 // but really is never defined.
 class __FlashStringHelper;
-#define FPSTR(pstr_pointer) (pstr_pointer)
-#define F(string_literal) (string_literal)
+#define FPSTR(pstr_pointer) (reinterpret_cast<const __FlashStringHelper *>(pstr_pointer))
+#define F(string_literal) (FPSTR(PSTR(string_literal)))
 
 // An inherited class for holding the result of a concatenation.  These
 // result objects are assumed to be writable by subsequent concatenations.

